### PR TITLE
Refactor HDFWriterModule for f142

### PR DESCRIPTION
**Pull request by dominikwerder**
### Description of work

- Factor classes into separate files
- Reduce repetitive dynamic type resolution code
- Add test

### Acceptance Criteria

- Tests should pass as before

### Unit Tests

- Added `tests/Schema_f142`

### CODE REVIEW

File: src/schemas/f142/FlatbufferReader.cpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#include "FlatbufferReader.h"
+
+namespace FileWriter {
+namespace Schemas {
+namespace f142 {
+
+FBUF const *get_fbuf(char const *data) { return GetLogData(data); }
+
+bool FlatbufferReader::verify(Msg const &msg) const {
+  auto veri = flatbuffers::Verifier((uint8_t *)msg.data(), msg.size());
```
Comment (matthew-d-jones) [inline src/schemas/f142/FlatbufferReader.cpp]: You can use `reinterpret_cast<const uint8_t *>(msg.data())`

File: src/schemas/f142/WriterArray.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+namespace f142 {
+
+template <typename DT, typename FV> class WriterArray : public WriterTypedBase {
+public:
+  WriterArray(hdf5::node::Group hdf_group, std::string const &source_name,
+              hsize_t ncols, Value fb_value_type_id, CollectiveQueue *cq);
+  WriterArray(hdf5::node::Group, std::string const &source_name, hsize_t ncols,
+              Value fb_value_type_id, CollectiveQueue *cq,
+              HDFIDStore *hdf_store);
+  ~WriterArray() override {}
```
Comment (matthew-d-jones) [inline src/schemas/f142/WriterArray.h]: Use `= default`

File: src/schemas/f142/WriterTypedBase.h [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+#include "../../h5.h"
+#include "Common.h"
+
+namespace FileWriter {
+namespace Schemas {
+namespace f142 {
+
+class WriterTypedBase {
+public:
+  virtual ~WriterTypedBase() {}
```
Comment (matthew-d-jones) [inline src/schemas/f142/WriterTypedBase.h]: Use `= default`

File: src/schemas/f142/f142_rw.cpp [hunk 4]
```diff
@@ -12,461 +15,212 @@ namespace f142 {
...
-    }
-    if (type == "double") {
-      return (R) new WA<double, ArrayDouble>(hg, s, array_size, vt, cq);
-    }
-  }
-  return (writer_typed_base *)nullptr;
+/// Helper struct to make branching on a found map entry more concise.
+template <typename T> struct FoundInMap {
+  FoundInMap() : Value(nullptr) {}
+  FoundInMap(T const &Value) : Value(&Value) {}
```
Comment (matthew-d-jones) [inline src/schemas/f142/f142_rw.cpp]: Mark `explicit`

File: src/schemas/f142/f142_rw.cpp [hunk 5]
```diff
@@ -12,461 +15,212 @@ namespace f142 {
...
-                                           cq, hdf_store);
-    }
-    if (type == "double") {
-      return (R) new WA<double, ArrayDouble>(hg, s, array_size,
-                                             Value::ArrayDouble, cq, hdf_store);
-    }
-  }
-  return (writer_typed_base *)nullptr;
+std::unique_ptr<WriterTypedBase>
+createWriterTypedBase(hdf5::node::Group HDFGroup, size_t ArraySize,
```
Comment (matthew-d-jones) [inline src/schemas/f142/f142_rw.cpp]: `HDFGroup` could be passed around as `const &`?

File: src/schemas/f142/f142_rw.h [hunk 6]
```diff
@@ -1,93 +1,182 @@
...
-  uptr<h5::h5d_chunked_1d<uint64_t>> ds_cue_index;
-  uptr<h5::h5d_chunked_1d<uint64_t>> ds_seq_data;
-  uptr<h5::h5d_chunked_1d<uint64_t>> ds_seq_fwd;
-  uptr<h5::h5d_chunked_1d<uint64_t>> ds_ts_data;
-  bool do_flush_always = false;
-  bool do_writer_forwarder_internal = false;
-  uint64_t total_written_bytes = 0;
-  uint64_t index_at_bytes = 0;
+  HDFWriterModule();
+  ~HDFWriterModule() {}
```
Comment (matthew-d-jones) [inline src/schemas/f142/f142_rw.h]: Mark `override` and use `= default`

File: src/tests/Schema_f142.cpp [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    WriterModule.TypeName = "float";
+    WriterModule.ArraySize = 0;
+    auto Group = File.root();
+    WriterModule.init_hdf(Group, "{}");
+    for (auto Value : Expected) {
+      auto Builder = makeValueFloat(Value);
+      auto Msg = FileWriter::Msg::owned(
+          reinterpret_cast<char *>(Builder->GetBufferPointer()),
+          Builder->GetSize());
+      WriterModule.write(std::move(Msg));
```
Comment (matthew-d-jones) [inline src/tests/Schema_f142.cpp]: `write` uses `Msg const &` so no move will actually happen here

File: src/schemas/f142/f142_rw.h [hunk 8]
```diff
@@ -1,93 +1,182 @@
...
+  /// Name of the HDF dataset
+  std::string Name;
+  /// Size of the HDF chunks in bytes
+  size_t ChunkBytes;
+  /// Size of the buffer which is used to optimize writing of small messages
+  size_t BufferSize;
+  /// Maximum size of a message which is considered for buffering
+  size_t BufferPacketMaxSize;
+  /// Helper
+  uptr<h5::h5d_chunked_1d<uint64_t>> *Ptr;
```
Comment (SkyToGround) [inline src/schemas/f142/f142_rw.h]: It is slightly dubious to have have `Ptr` be a pointer to a `std::unique_ptr`. Although (after checking more closely) this code appears to be correct, it is dangerous. In my opinion `Ptr` should be either a `std::shared_ptr` or a raw pointer. Regardless, it should definitely be initialised to `nullptr`.

Comment (dominikwerder) [reply src/schemas/f142/f142_rw.h]: True, I'll fix that.

**Pull request merged**